File: test/unit/lua/gc_optimized_map_spec.lua
local GCOptimizedMap = require("tstl-gc-optimized-collections.lua.gc_optimized_map").GCOptimizedMap

describe("GCOptimizedMap", function()
    describe("constructor", function()
        local my_map = GCOptimizedMap.new({1, "hello"}, {2, "hey"})

        it("should have size 2", function()
            assert.is.equal(2, my_map.size)
        end)

        it("should have the elements", function()
            assert.is.equal("hello", my_map:get(1))
            assert.is.equal("hey", my_map:get(2))
        end)
    end)

    describe("clear", function()
        local my_map = GCOptimizedMap.new({2, 5}, {3, 89})

        my_map:clear()

        it("should have size 0", function()
            assert.is.equal(my_map.size, 0)
        end)

        it("should not have items", function()
            assert.is_nil(my_map:get(2))
            assert.is_nil(my_map:get(3))
        end)
    end)

    describe("delete", function()
        local my_map = GCOptimizedMap.new({4, 2}, {1, 2}, {7, 3})

        my_map:delete(1)

        it("removes the entry", function()
            assert.is_nil(my_map:get(1))
        end)
    end)

    describe("entries", function()
        it("iterates through the entries", function()
            local my_map = GCOptimizedMap.new({1, 4}, {2, 3}, {5, 6})

            local key_sum = 0
            local value_sum = 0
            for k, v in my_map:entries() do
                key_sum = key_sum + k
                value_sum = value_sum + v
            end

            assert.is.equal(8, key_sum)
            assert.is.equal(13, value_sum)
        end)
    end)

    describe("get", function()
        local my_map = GCOptimizedMap.new({1, 4}, {2, 3}, {5, 6})

        it("retrieves the correct value", function()
            assert.is.equal(6, my_map:get(5))
        end)
    end)

    describe("has", function()
        local my_map = GCOptimizedMap.new({1, 4}, {2, 3}, {5, 6})

        describe("the map has the key", function()
            it("returns true", function()
                assert.is_true(my_map:has(2))
            end)
        end)

        describe("the map does not have the key", function()
            it("returns false", function()
                assert.is_false(my_map:has(8))
            end)
        end)
    end)

    describe("set", function()
        local my_map = GCOptimizedMap.new()

        my_map:set(5, 0)
        my_map:set(2, 6)

        it("sets the key value pair", function()
            assert.is.equal(0, my_map:get(5))
            assert.is.equal(6, my_map:get(2))
        end)
    end)

    describe("size", function()
        local my_map = GCOptimizedMap.new({1, 4}, {2, 3}, {5, 6})

        it("returns the amount of entries", function()
            assert.is.equal(3, my_map.size)
        end)
    end)
end)

File: test/unit/ts/gc_optimized_map_spec.ts
import { GCOptimizedMap } from "../../../tstl-gc-optimized-collections/init";

describe("GCOptimizedMap", () => {
    describe("Empty", () => {
        it("should be empty", () => {
            expect(GCOptimizedMap.Empty.size).toBe(0);
        });
    });

    describe("constructor", () => {
        describe("no argument", () => {
            const myMap = new GCOptimizedMap();

            it("should have size 0", () => {
                expect(myMap.size).toBe(0);
            });
        });

        describe("iterable argument", () => {
            const myMap = new GCOptimizedMap(["a", "c"],["b", "d"]);

            it("should have size 2", () => {
                expect(myMap.size).toBe(2);
            });

            it("should have element a", () => {
                expect(myMap.has("a")).toBeTruthy();
            });

            it("should have element b", () => {
                expect(myMap.has("b")).toBeTruthy();
            });
        });
    });

    describe("clear", () => {
        const myMap = new GCOptimizedMap<string, string>(["a","c"], ["b","d"]);

        myMap.clear();

        it("should have size 0", () => {
            expect(myMap.size).toBe(0);
        });

        it("should not have first element", () => {
            expect(myMap.has("a")).toBeFalsy();
        });

        it("should not have second element", () => {
            expect(myMap.has("b")).toBeFalsy();
        });
    });

    describe("delete", () => {
        const myMap = new GCOptimizedMap(["a", "c"],["b", "d"]);

        myMap.delete("a");

        it("should not have element a", () => {
            expect(myMap.has("a")).toBeFalsy();
        });

        it("should have element b", () => {
            expect(myMap.has("b")).toBeTruthy();
        });
    });

    describe("forEach", () => {
        const myMap = new GCOptimizedMap(["a", 2], ["b", 3], ["c", 4]);
        let count = 0;
        for (const [k, v] of myMap.entries()) {
            count += v;
        }

        it("iterates over all entries", () => {
            expect(count).toBe(9);
        });
    });

    describe("get", () => {
        const myMap = new GCOptimizedMap(["a", "c"], ["b", "d"]);

        describe("the map has the key", () => {
            it("gets the value from the key", () => {
                expect(myMap.get("a")).toBe("c");
            });
        });

        describe("the map does not have the key", () => {
            it("returns undefined", () => {
                expect(myMap.get("c")).toBeUndefined();
            });
        });
    });

    describe("has", () => {
        const myMap = new GCOptimizedMap(["a", "c"]);

        describe("the map has the key", () => {
            it("returns true", () => {
                expect(myMap.has("a")).toBeTruthy();
            });
        });

        describe("the map does not have the key", () => {
            it("returns false", () => {
                expect(myMap.has("c")).toBeFalsy();
            });
        });
    });

    describe("set", () => {
        const myMap = new GCOptimizedMap();

        myMap.set("a", 5);

        it("should have key", () => {
            expect(myMap.has("a")).toBeTruthy();
        });

        it("should have value", () => {
            expect(myMap.get("a")).toBe(5);
        });
    });

    describe("size", () => {
        describe("the map is empty", () => {
            const m = new GCOptimizedMap();

            it("returns 0", () => {
                expect(m.size).toBe(0);
            });
        });

        describe("the map has a value set", () => {
            const m = new GCOptimizedMap();

            m.set(1, 3);

            it("returns 1", () => {
                expect(m.size).toBe(1);
            });
        });

        describe("the map is constructed with 2 entries", () => {
            const m = new GCOptimizedMap([1,2],[3,4]);

            it("returns 2", () => {
                expect(m.size).toBe(2);
            });
        });

        describe("the map is cleared", () => {
            const m = new GCOptimizedMap([1,2],[3,4]);
            m.clear();

            it("returns 0", () => {
                expect(m.size).toBe(0);
            });
        });

        describe("the map has 2 entries and 1 is deleted", () => {
            const m = new GCOptimizedMap([1,2], [3,4]);
            m.delete(1);

            it("returns 1", () => {
                expect(m.size).toBe(1);
            });
        });
    });

    describe("entries", () => {
        const m = new GCOptimizedMap([3, 2], [1, 4], [2, 6]);

        it("iterates through the entries", () => {
            let key_sum = 0;
            let value_sum = 0;
            for (const [k, v] of m.entries()) {
                key_sum += k;
                value_sum += v;
            }
            expect(key_sum).toBe(6);
            expect(value_sum).toBe(12);
        });
    });
});
